### PR TITLE
Make manifest a nullable field on file repos and publications

### PR DESCRIPTION
**Pull request by daviddavis**
Goes with https://github.com/pulp/pulp_file/pull/838

### CODE REVIEW

**Review by ggainey — CHANGES_REQUESTED** [review]
Hey @daviddavis  - can we add a test for this?

**Comment by daviddavis** [comment]
I tried running the tests but I am hitting this error:

```
$ make test
pytest -v tests
Traceback (most recent call last):
  File "/usr/local/lib/pulp/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/config/__init__.py", line 187, in console_main
    code = main()
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/config/__init__.py", line 145, in main
    config = _prepareconfig(args, plugins)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/config/__init__.py", line 324, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pluggy/_callers.py", line 55, in _multicall
    gen.send(outcome)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/helpconfig.py", line 102, in pytest_cmdline_parse
    config: Config = outcome.get_result()
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/config/__init__.py", line 1016, in pytest_cmdline_parse
    self.parse(args)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/config/__init__.py", line 1304, in parse
    self._preparse(args, addopts=addopts)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/config/__init__.py", line 1187, in _preparse
    self.pluginmanager.load_setuptools_entrypoints("pytest11")
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pluggy/_manager.py", line 287, in load_setuptools_entrypoints
    plugin = ep.load()
  File "/usr/lib64/python3.10/importlib/metadata/__init__.py", line 171, in load
    module = import_module(match.group('module'))
  File "/usr/lib64/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/_pytest/assertion/rewrite.py", line 168, in exec_module
    exec(co, module.__dict__)
  File "/home/vagrant/devel/pulpcore/pulpcore/tests/functional/__init__.py", line 5, in <module>
    from pulp_smash.config import get_config
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pulp_smash/config.py", line 21, in <module>
    from pulpcore.client.pulpcore import Configuration
ImportError: cannot import name 'Configuration' from 'pulpcore.client.pulpcore' (unknown location)
make: *** [Makefile:26: test] Error 1
```

Any idea?

**Comment by daviddavis** [comment]
Also, I am trying to follow [the contribution guide instructions for testing](https://github.com/pulp/pulp-cli/blob/main/docs/contributing.md#testing) but it seems woefully incomplete. Is there something else I should be using?

**Comment by mdellweg** [comment]
This looks like the Pulp-smash pytest plugin was loaded, but not configured. The cli tests don't use it...

**Comment by daviddavis** [comment]
I tried uninstalling pulp-smash but then I get an error that it can’t import it. So weird that somehow pulpcore code is getting executed.

**Comment by mdellweg** [comment]
It's probably the other way around. Pulp-smash is a pytest plugin and loads without being asked. Try uninstalling it.

**Review by mdellweg — APPROVED** [review]

**Review by ggainey — DISMISSED** [review]
The code looks fine - but pulp/pulp_file#837 isn't backported, so all the non-nightly tests fail because the images they're based  have only released versions of pulp_file, where you can't PATCH the manifest to null. That specific test needs its own min-version check.

File: pulpcore/cli/file/context.py [hunk 1]
```diff
@@ -59,6 +59,7 @@ class PulpFilePublicationContext(PulpEntityContext):
     HREF = "file_file_publication_href"
     ID_PREFIX = "publications_file_file"
     CAPABILITIES = {"roles": [PluginRequirement("file", min="1.11.0")]}
+    NULLABLES = {"manifest"}
 
     def preprocess_entity(self, body: EntityDefinition, partial: bool = False) -> EntityDefinition:
         body = super().preprocess_entity(body, partial=partial)
```
Comment (mdellweg) [inline pulpcore/cli/file/context.py:65]: We could add an additional client side version check on `manifest is None` here to have validation before sending the body to the server.

Comment (daviddavis) [reply pulpcore/cli/file/context.py:65]: It didn't seem necessary IMO. Users get a user friendly message when they try to set manifest to null for older versions of pulp_file:

```
$ pulp file repository create --name test --manifest ""
Error: {"manifest":["This field may not be null."]}
```

Comment (mdellweg) [reply pulpcore/cli/file/context.py:65]: I agree. This was just recalling @ggainey 's mantra: "Whatever we can catch before bugging the server..."

**Review by mdellweg — APPROVED** [review]

**Pull request merged**